cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -1565,6 +1565,103 @@ data MultiSigScript = RequireSignature (Hash PaymentKey)
+
+instance FromJSON MultiSigScript where
+  parseJSON = Aeson.withObject "MultiSigScript" $ \obj -> do
+    -- Parse "all" multisig objects
+    allReqHashes <- case HMS.lookup "all" obj of
+                       Just (Array vecAllReqHashes) ->
+                         return . Just . RequireAllOf $ gatherSignatures vecAllReqHashes
+                       _ -> return Nothing
+
+    -- Parse "any" multisig objects
+    anyReqHashes <- case HMS.lookup "any" obj of
+                       Just (Array vecAnyReqHashes) ->
+                         return . Just . RequireAnyOf $ gatherSignatures vecAnyReqHashes
+                       _ -> return Nothing
+
+    -- Parse "mofn" multisig objects
+    mofN <- return $ filterMofN obj
+
+    case (allReqHashes, anyReqHashes, mofN) of
+      (Nothing, Nothing, Nothing) -> fail "Script does not contain \"all\", \"any\" \
+                                          \or \"mofn\" JSON objects."
+      (Just allMss, Nothing, Nothing) -> return allMss
+      (Nothing, Just anyMss, Nothing) -> return anyMss
+      (Nothing, Nothing, Just mOfnMss) -> return mOfnMss
+      (_, _, _) -> fail "Please specify only one type of JSON multisig object."
+
+gatherSignatures :: Vector Value -> [MultiSigScript]
+gatherSignatures anyList =
+  let mPaymentKeyHashes = Vector.map (\val -> RequireSignature <$> filterValue val) anyList
+  in catMaybes $ Vector.toList mPaymentKeyHashes
+ where
+  filterValue :: Value -> Maybe (Hash PaymentKey)
+  filterValue (String hpk) = Just $ convertToHash hpk
+  filterValue _ = Nothing
+
+  convertToHash :: Text -> Hash PaymentKey
+  convertToHash txt = case deserialiseFromRawBytesHex (AsHash AsPaymentKey) $ Text.encodeUtf8 txt of
+                        Just payKeyHash -> payKeyHash
+                        Nothing -> panic $ "Error deserialising payment key hash: " <> txt
+
+type Parser = MParsec.Parsec Void String
+
+filterMofN :: HMS.HashMap Text Value -> Maybe MultiSigScript
+filterMofN obj = do
+  let mofn = map onlyMofN $ allPotentialScripts obj
+  case mofn of
+    -- "mofn"
+    [Just (m, of', n)] ->
+      case HMS.lookup (Text.pack $ show m ++ of' ++ show n) obj of
+        Just (Array mHashed) -> case n == Vector.length mHashed of
+                                  True -> Just . RequireMOf m
+                                            $ gatherSignatures mHashed
+                                  False -> panic "mofn error: n is not equal to the number of payment key hashes provided."
+        _ -> Nothing
+    -- "any", "all"
+    _ -> Nothing
+ where
+   allPotentialScripts :: HMS.HashMap Text Value -> [ParsedScript]
+   allPotentialScripts obj' = catMaybes . map (parseMaybe scriptParser . Text.unpack)
+                                $ HMS.keys obj'
+
+   onlyMofN :: ParsedScript -> Maybe (Int, String, Int)
+   onlyMofN (All _) = Nothing
+   onlyMofN (Any _) = Nothing
+   onlyMofN (MofN mon) = Just mon
+
+
+data ParsedScript = All String | Any String | MofN (Int, String, Int)
```

Comment:
I think the `String` in `MofN (Int, String, Int)` is redundant because it always has the same value.  Can delete without loss of functionality.

Reply:
Yep that's a good point.